e2e/nomostest/gitproviders/bitbucket.go

```diff
@@ -94,6 +95,11 @@ func (b *BitbucketClient) CreateRepository(localName string) (string, error) {
 		repoName = repoName[:repoNameMaxLength]
 	}
 
+	// last character should be alphanumeric
+	if !regexp.MustCompile(`^[a-zA-Z0-9]*$`).MatchString(repoName[len(repoName)-1:]) {
+		repoName = repoName[:len(repoName)-1]
```

Comment:
what repo names end in non-alphanumeric characters???

Reply:
can we just add validation here that fails the test and then go fix the test to not have a crazy name?

Reply:
or it it just that it gets truncated in the middle of the name that it might end in a hyphen?

Reply:
To create a unique repository name, we generate a random string using `uuid.NewRandom()`, append it to the provided r*sync name, and truncate it to keep it within maximum length. The truncated name can sometimes ends with a dash ('-') character that will cause the bitbucket create repo API to throw an error

Reply:
Here's a better idea, taken from a similar problem we had with GKE subnet names in firewall rules:

1. Use a shorter max-length for the localName (like max-8, to allow at least 7 UUID characters), and truncate that first. This leaves enough characters to append a minimum length uuid suffix, which ensures that extra long names don't become non-unique after truncation.
2. Strip all the hyphens out of the UUID before appending it to the localName. This ensures the string will never end in a hyphen.
3. Do another max-length truncation after appending. This ensures use of the longest possible UUID, depending on localName length.

Reply:
localName always start with the namespace name, so typically RootSync resource name will look like `config-management-system-root-sync`. Meanwhile, UUID is always 36 characters, 32 if we stripped all the hyphens. The current max length for repo name is set to 62, so maybe we can allocate 30 characters for localName and 32 for UUID (all stripped of hyphens)? With that, typical repoName will look something like: `configmanagementsystemrootsync43b555fe7764f76bbb1497f2492d8ed`

Reply:
updated to follow Karl's solution in https://github.com/GoogleContainerTools/kpt-config-sync/pull/511#discussion_r1156511615